In [3]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib as plt

#renamed file to "data.csv" for convenience
data = pd.read_csv("data.csv")

Let's explore the data...

In [28]:
max(data['order_amount'])

704000

This number seems to high to not be a mistake of some kind... let's explore the order_amount column

In [16]:
order_amount = data.sort_values("order_amount", ascending = False)
order_amount.head(25)

order_id  shop_id  user_id  order_amount  total_items payment_method  \
2153      2154       42      607        704000         2000    credit_card   
3332      3333       42      607        704000         2000    credit_card   
520        521       42      607        704000         2000    credit_card   
1602      1603       42      607        704000         2000    credit_card   
60          61       42      607        704000         2000    credit_card   
2835      2836       42      607        704000         2000    credit_card   
4646      4647       42      607        704000         2000    credit_card   
2297      2298       42      607        704000         2000    credit_card   
1436      1437       42      607        704000         2000    credit_card   
4882      4883       42      607        704000         2000    credit_card   
4056      4057       42      607        704000         2000    credit_card   
15          16       42      607        704000         2000    credit_card   
1104      1105       42      607        704000         2000    credit_card   
1562      1563       42      607        704000         2000    credit_card   
2969      2970       42      607        704000         2000    credit_card   
4868      4869       42      607        704000         2000    credit_card   
1362      1363       42      607        704000         2000    credit_card   
691        692       78      878        154350            6          debit   
2492      2493       78      834        102900            4          debit   
3724      3725       78      766         77175            3    credit_card   
4420      4421       78      969         77175            3          debit   
4192      4193       78      787         77175            3    credit_card   
3403      3404       78      928         77175            3          debit   
2690      2691       78      962         77175            3          debit   
2564      2565       78      915         77175            3          debit   

               created_at  
2153   2017-03-12 4:00:00  
3332   2017-03-24 4:00:00  
520    2017-03-02 4:00:00  
1602   2017-03-17 4:00:00  
60     2017-03-04 4:00:00  
2835   2017-03-28 4:00:00  
4646   2017-03-02 4:00:00  
2297   2017-03-07 4:00:00  
1436   2017-03-11 4:00:00  
4882   2017-03-25 4:00:00  
4056   2017-03-28 4:00:00  
15     2017-03-07 4:00:00  
1104   2017-03-24 4:00:00  
1562   2017-03-19 4:00:00  
2969   2017-03-28 4:00:00  
4868   2017-03-22 4:00:00  
1362   2017-03-15 4:00:00  
691   2017-03-27 22:51:43  
2492   2017-03-04 4:37:34  
3724  2017-03-16 14:13:26  
4420  2017-03-09 15:21:35  
4192   2017-03-18 9:25:32  
3403   2017-03-16 9:45:05  
2690   2017-03-22 7:33:25  
2564   2017-03-25 1:19:35

Question 1a)

As shown above, there seems to be an issue with shop_id = 42 and shop_id = 78, and their order amount number. In order to fix this issue, I would propose removing the rows that correspond to shop_id = 42. It seems as if every data entry for shop_id = 42 is the same, so these rows wouldn't be helpful for our analysis. The only different value is the order_id, and every other value is the same. As far as shop_id = 78, it is possible that the order_amount was mistakenly entered in cents, so that can be converted by dividing each order_amount with shop_id = 78 by 100. This is done below.

In [30]:
converted = data[data['shop_id'] == 78]
converted['order_amount'] = converted['order_amount'] / 100
clean_data = data[(data['shop_id'] != 42) & (data['shop_id'] != 78)]
clean = pd.concat([clean_data, converted])

/tmp/ipykernel_44/608747781.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  converted['order_amount'] = converted['order_amount'] / 100


Question 1b) and 1c)

Lets take a look at the different metrics we can use, now that we've cleaned the data.

In [31]:
clean['order_amount'].describe()

count    4949.000000
mean      301.940190
std       158.329795
min        90.000000
25%       163.000000
50%       284.000000
75%       387.000000
max      1543.500000
Name: order_amount, dtype: float64

Mean doesn't seem like the best choice, as it seems to be much higher than the 50th percentile, and is influenced heavily by the outliers. In this case, I would instead prefer to report median as the best metric because it isn't affected by the large outliers. This report would lead to a value of 284, which is much more realistic than the previous report.

Question 2

In [32]:
# SQL CODE
# SELECT ShipperName, COUNT(ShipperName) OrdersShipped FROM (
# SELECT *
# FROM Orders o
# LEFT JOIN Shippers s
# ON s.ShipperID = o.ShipperID)
# GROUP BY ShipperName
# HAVING ShipperName = 'Speedy Express'
# ;

#Answer is 54 orders


In [1]:
# SELECT * FROM(
# (SELECT EmployeeID, COUNT(*) NumOrders FROM [Orders]
# GROUP BY EmployeeID) e
# LEFT JOIN (SELECT EmployeeID, LastName FROM Employees) a
# ON e.EmployeeID = a.EmployeeID)
# ORDER BY NumOrders DESC
# LIMIT 1

#Answer is Peacock with 40 orders

In [2]:
# SELECT ProductName, Country, TotalQuantity FROM (
# SELECT ProductId, Country, SUM(Quantity) TotalQuantity FROM OrderDetails od LEFT JOIN (
# SELECT o.OrderID, o.CustomerID, c.Country FROM Orders o
# LEFT JOIN (SELECT CustomerId, Country FROM Customers) c
# ON o.CustomerID = c.CustomerID) custom
# ON od.OrderID = custom.OrderID
# GROUP BY Country, ProductID
# HAVING Country = 'Germany') germany
# LEFT JOIN (SELECT ProductId, ProductName FROM Products) p
# ON germany.ProductId = p.ProductId ORDER BY TotalQuantity DESC LIMIT 1

#Answer is Boston Crab Meat